lib/Sema/TypeCheckStorage.cpp

```diff
@@ -2362,6 +2415,105 @@ PropertyWrapperMutabilityRequest::evaluate(Evaluator &,
   return result;
 }
 
+llvm::Expected<Optional<PropertyWrapperLValueness>>
+PropertyWrapperLValuenessRequest::evaluate(Evaluator &,
+                                           VarDecl *var) const {
+    VarDecl *VD = var;
+    unsigned numWrappers = var->getAttachedPropertyWrappers().size();
+    bool isProjectedValue = false;
+    if (numWrappers < 1) {
+      VD = var->getOriginalWrappedProperty(
+        PropertyWrapperSynthesizedPropertyKind::StorageWrapper);
+      numWrappers = 1; // Can't compose projected values
+      isProjectedValue = true;
+    }
+
+    if (!VD)
+      return None;
+
+    auto varMember = isProjectedValue
+      ? &PropertyWrapperTypeInfo::projectedValueVar
+      : &PropertyWrapperTypeInfo::valueVar;
+
+    // Calling the getter (or setter) on the nth property wrapper in the chain
+    // is done as follows:
+    //  1. call the getter on the (n-1)th property wrapper instance to get the
+    //     nth property wrapper instance
+    //  2. call the getter (or setter) on the nth property wrapper instance
+    //  3. if (2) is a mutating access, call the setter on the (n-1)th property
+    //     wrapper instance to write back the mutated value
+
+    // Below, we determine which of these property wrapper instances need to be
+    // accessed mutating-ly, and therefore should be l-values.
+
+    // The variables used are:
+    //
+    //    - prevWrappersMutabilityForGet:
+    //
+    //      The mutability needed for the first (n-1) wrapper instances to be
+    //      able to call the getter on the (n-1)th instance, for step (1) above
+    //
+    //    - prevWrappersMutabilityForGetAndSet:
+    //
+    //      The mutability needed for the first (n-1) wrapper instances to be
+    //      able to call both the getter and setter on the (n-1)th instance, for
+    //      steps (1) and (3) above
+    //
+    //    - mutability:
+    //
+    //      The mutability needed for calling the getter / setter on the
+    //      nth wrapper instance, for step (2) above.
+
+    llvm::SmallVector<PropertyWrapperMutability::Value, 1>
+      prevWrappersMutabilityForGet, prevWrappersMutabilityForGetAndSet;
+    PropertyWrapperMutability mutability;
+
+    auto firstWrapperInfo = VD->getAttachedPropertyWrapperTypeInfo(0);
+    mutability.Getter = getGetterMutatingness(firstWrapperInfo.*varMember);
+    mutability.Setter = getSetterMutatingness(firstWrapperInfo.*varMember,
+                                var->getInnermostDeclContext());
+
+    for (unsigned i : range(1, numWrappers)) {
```

Comment:
I think this bit of code would make more sense if you first compute the `PropertyWrapperMutability` for each composed property wrapper, and then compute the mutability sequence for the last get/set from that.

Reply:
Do you mean something like: Given Swift code like `@A @B @C var member: Int`, compute `PropertyWrapperMutability` for `@A`, then `@A @B`, then `@A @B @C`, and derive the lvalueness based on those results?

Reply:
This isn't what I had in mind - I'm going to revisit your code here to deepen my understanding and figure out if what I had in mind is possible